pkg/core/chain/chain.go

```diff
@@ -157,134 +139,87 @@ func New(ctx context.Context, db database.DB, eventBus *eventbus.EventBus, rpcBu
 	return chain, nil
 }
 
-// SetupConsensus adds the missing fields on the Chain which need to be populated
-// by the user. Once the fields are populated, consensus is started.
-// NOTE: we need locking here so that there is no race condition between
-// the synchronization routine trying to figure out if there is a loop
-// available, and the setting of these fields.
-func (c *Chain) SetupConsensus(pk keys.PublicKey, l *loop.Consensus) {
-	c.lock.Lock()
-	c.pubKey = &pk
-	c.loop = l
-	c.lock.Unlock()
-
-	// If we are on genesis and our consensus is being set up, it means
-	// the network has just bootstrapped. So, we need to start the consensus,
-	// and set up the interrupt.
-	if c.tip.Header.Height == 0 {
-		go func() {
-			if err := c.startConsensus(); err != nil {
-				log.WithError(err).Error("consensus failure")
-			}
-		}()
-	}
+func (c *Chain) currentHeight() uint64 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.tip.Header.Height
 }
```

Comment:
Because using the DB would be more racey than fetching it like this - we can lock during the acceptance procedure, so that we can guarantee that the caller will get the most updated tip at any point. If we use the DB, this synchronization is lost and may cause unwanted behaviour in the `Synchronizer`.

Reply:
The DB cannot be racey by definition. The only way using the DB can lead to inconsistencies is when the stored data is  updated while some other process is reading it. To avoid this, we should do all writes in a single DB transaction, which is inherently threadsafe. The fact that using the DB can lead to race conditions is a huge smell and we should not mask it with caching in the first place.

Reply:
TBC, when we are synchronizing the node should do nothing else apart from republishing stuff. Therefore I am not really sure who would actually use the tip information when we are synching

Reply:
That's not what I meant, let me explain a scenario:

The chain is on tip height 5, and we are in syncing mode. The synchronizer receives block 6 from the network, so it checks the tip height from the DB (`FetchState`), notices that it is one ahead of the tip, and so calls `processSyncBlock`. This starts the acceptance procedure. Now imagine, that while the state transition is executed, and the tip has not yet been updated, block 7 arrives from the network. The synchronizer checks the state, and notices it is 2 ahead of the tip. So, it puts the block in the `sequencer`. When the block acceptance procedure is finished, the tip is updated to height 6. The next block to come in will be any block after height 7, and thus all blocks will just be put in the sequencer and an acceptance procedure is never initiated, because we already stored the successive block.

There are a few ways to mitigate such issues, but I don't know if we can 100% cover this scenario without introducing some sort of locking. We could check the sequencer every time a block is received in the synchronizer, but this may cause some stalls near the end of a sync, and cause nodes to stay behind for longer than is necessary. Another would be to check the sequencer after finishing block acceptance, but there is also no guarantee that we time it correctly in this case.

Unless I've made a mistake in implementing this function, I don't see how using the DB directly wouldn't cause at least some kind of discomfort when handling this scenario.

Reply:
I told you in the previous comment. By using DB transaction you inherently introduce thread-safety. If you create a DB Tx at the beginning of the block acceptance routine and release it (either by rolling back if the verification is unsuccesful, or by updating the data) when the routine is done, you apply an intrinsic locking mechanism built in on the DBs. This is pretty standard on SQL DBs, especially when updating multiple tables at once. The only thing I am not sure about, it is whether LevelDB supports transactions. If this is not available, you'd be right to perform the locking manually on the caching side.

Reply:
Ah, I see. I missed that. I will add it

Reply:
As it turns out, LevelDB does not support atomic transactions. The alternative mechanism it uses (`WriteBatch`) is not suitable for locking the tip data. Furthermore, I am not even sure locking would suffice to cover the scenario you describe. In the previous alternative design, the tip was handled by a specific component outside of the chain. One option would be to revive that wrapper, make it threadsafe, encapsulate the DB operations there and enrich it with a transaction-like interface, if it would help with the current scenario